Add support for built-in functions

mirryi opened the pull request:
This ports over the mechanisms for supporting builtin functions from the `strings3` branch. Builtin functions are declared in `hazelcore/Builtins.re` and implemented (in Reason/OCaml, not Hazel) in `hazelcore/dynamics/Evaluator.re` (see `evaluate_ap_builtin`).

To start, `int_of_float` and `float_of_int` have been added.

mirryi committed:
Add Builtins.re with some basic builtin functions

mirryi committed:
Implement evaluation for builtins

mirryi committed:
Replace subst_var_list with List.map

mirryi committed:
Implement ui components of ApBuiltin

mirryi committed:
Set empty Context to builtins

mirryi committed:
Substitute builtin function name with wrapping lam

mirryi committed:
Fix builtin function substitution

mirryi committed:
Add todo notice for builtin substitution

cyrus- commented on src/hazelcore/Contexts.re:
```diff
@@ -1,6 +1,6 @@
 [@deriving sexp]
 type t = (VarCtx.t, PaletteCtx.t);
-let empty = (VarCtx.empty, PaletteCtx.empty);
+let empty = (Builtins.ctx, PaletteCtx.empty);
```
let's call this `initial` since its not actually empty

cyrus- commented on src/hazelcore/dynamics/Evaluator.re:
```diff
@@ -816,4 +829,31 @@ and evaluate_case =
+  }
+and evaluate_ap_builtin = (ident: string, args: list(DHExp.t)): result => {
+  switch (ident) {
+  | "int_of_float" =>
```
let's decentralize this logic into a lookup table in the Builtins module rather than with a switch here, so new builtins only require updating one file ideally

mirryi replied:
I tried doing this initially, but I got circular module dependency errors (the table in `Builtins` would rely on `Evaluator.result`, and `Evaluator` has to look up in `Builtins`. One solution would be to move `Evaluator.result` into its own module?

cyrus- replied:
Yeah moving Evaluator.result into EvaluatorResult.t makes sense.

cyrus- replied:
actually calling it EvaluationResult.t is better so that we can reuse it even when we have a compiler

mirryi replied:
Cool, I'll do that

cyrus- commented on src/hazelweb/layout/DHDoc_Exp.re:
```diff
@@ -188,6 +189,19 @@ let rec mk =
+      | ApBuiltin(x, args) =>
+        switch (args) {
+        // TODO: How to handle tl?
+        | [hd, ..._] =>
```
arguments are missing here -- best way to handle it is probably to call into the Ap case

mirryi committed:
Rename Contexts.empty to Contexts.initial

mirryi committed:
Move Evaluator.result to own module

mirryi committed:
Move builtin implementations to Builtins.re

mirryi committed:
Use Ap case for ApBuiltin in DHDoc_Exp.mk

mirryi committed:
Merge branch 'dev' into builtin-fns

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -1,6 +1,46 @@
+module Impls = {
```
Let's create a single mapping from the builtin name to both its type and its evaluation behavior, so that there is one localized place where each builtin is implemented.

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -1,6 +1,46 @@
+module Impls = {
+  open EvaluatorResult;
+
+  let int_of_float = (ident, args, evaluate) =>
```
It might be easier to define a functor around the entire Builtins module rather than taking in the evaluator in each function.

mirryi replied:
How can we create a functor for a file, as opposed to a nested module inside the file module?

mirryi replied:
`evaluate` isn't really relevant to other places that use `Builtins` (such as `Contexts.re`, which uses `Builtins.ctx`), so I think that a functor is perhaps not the best approach.

cyrus- replied:
You can define an FBuiltins.F functor that takes the evaluator as input, then define Builtins.re as `include FBuiltins(Evaluator)` and use Builtins normally from there. That breaks up the circularity.

mirryi replied:
Hmm, I tried some different things but couldn't figure out how to do what you mean and avoid a circular dependency. If Builtins has `include FBuiltins(Evaluator)`, and then Evaluator calls Builtins, there would still be a circular dependency?

cyrus- commented on src/hazelcore/dynamics/Evaluator.re:
```diff
@@ -528,9 +526,15 @@ let eval_bin_float_op =
+  | BoundVar(x) =>
+    switch (Builtins.lookup_type(x)) {
+    | Some(ty) =>
+      // TODO: Assumes all builtins are functions
```
this approach also means we can't shadow built-in functions, which I think would be fine to allow. 

I think a better approach is to insert all of the built-ins as let bindings wrapping ApBuiltin at the top of the program after elaboration is complete.

mirryi replied:
I added a new `elab` function to `Elaborator_Exp.re` that we call from the outside instead of `syn_elab` directly. In there I insert the builtins at the top. Is that what you were thinking?

mirryi committed:
Reorganize Builtins to require registering new builtin fn in one place

mirryi committed:
Allow shadowing of builtins by inserting at beginning after elaboration

mirryi committed:
Merge branch 'dev' into builtin-fns

mirryi committed:
Add mk_one_arg helper for creating one arg builtins

mirryi committed:
Automatically evaluate d1 in mk_one_arg

mirryi committed:
Merge branch 'dev' into builtin-fns

mirryi committed:
Merge branch 'dev' into builtin-fns

mirryi committed:
Merge branch 'dev' into builtin-fns

mirryi committed:
Format Builtins module

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,55 @@
+  let mk_one_arg = (f, ident, args, evaluate) => {
+    let e = DHExp.ApBuiltin(ident, args);
+    switch (args) {
+    | [] => Indet(e)
```
should raise an exception if supplied an incorrect number of arguments, not just return Indet

mirryi replied:
Fixed.

cyrus- committed:
small code clarity tweaks

cyrus- commented on src/hazelcore/dynamics/Elaborator_Exp.re:
```diff
@@ -833,3 +833,22 @@ and ana_elab_rule =
+        (d', (x, _)) =>
+          DHExp.Let(
+            Var(x),
+            Lam(Var("x"), ty, ApBuiltin(x, [BoundVar("x")])),
```
how would this handle builtins with more than one argument, if curried?

mirryi committed:
Raise exception when builtin evaluated with incorrect number of arguments

mirryi committed:
Overhaul definition of builtins to make constants and currying work

mirryi committed:
Clean up and some documentation

mirryi commented:
I reworked much of it so that after elaboration, the let bindings inserted around the expression have a structure that mirrors how it would be if the functions were actually written in Hazel; currying works now (see the `mod` example I added). Defining new built-ins in the `Builtins` module is still pretty much identical to before.

We can also define built-in non-function constants now. I added a `PI` constant for reference.

mirryi committed:
Add error handling for when mod when modulus is 0

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,183 @@
+    (
+      /* args: */ args,
+      /* evaluate: */ DHExp.t => EvaluatorResult.t,
+      /* exp: */ DHExp.t
```
why are we passing in the initial unevaluated expression? to support Indet? if so, it isn't quite right -- the indet expression should have fully evaluated args even if it can't be further reduced, not left entirely unevaluated...

mirryi replied:
Not sure why I did it like this originally... should be fixed now, I think.

mirryi committed:
Fix Indet evaluation for built-ins

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,181 @@
+);
+
+[@deriving sexp]
+type t = (Var.t, HTyp.t, form);
```
let's just put the identifier, type, eval function and elab all in one record with field names instead of broken out like this

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,181 @@
+module Impl = {
+  [@deriving sexp]
+  type f =
+    (/* args: */ args, /* evaluate: */ DHExp.t => EvaluatorResult.t) =>
```
probably good to define an evaluator type alias for readability

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,181 @@
+
+  /* int_of_float implementation. */
+  let int_of_float = (ident, d1') =>
+    switch (d1') {
```
Let's use `r` variables for evaluator results and keep `d` only for DHExps so that its easier to read

cyrus- reviewed (changes requested):
OK looks much better now. A few last tweaks!

mirryi committed:
Use r* variables for ElaboratorResult.t values

mirryi committed:
Clean up Builtin types

mirryi commented:
Implemented the suggestions. Everything look good now?

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,190 @@
+type elab = DHExp.t;
+
+[@deriving sexp]
+type t = {
```
Naming is maybe still a bit off. This type should probably be called Builtin.t. Then you can define a Builtins.t module which actually defines the specific builtins that we are defining for Hazel right now.

mirryi replied:
Alright, that sounds good. `mod` and `PI` I added for making sure currying and constants work. Do we want to keep those? And what other built-ins might be wanted for now?

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,190 @@
+  type evaluate = DHExp.t => EvaluatorResult.t;
+
+  [@deriving sexp]
+  type f = (/* args: */ args, /* evaluate: */ evaluate) => EvaluatorResult.t;
```
these comments can be type annotations now I think?

mirryi committed:
Move Builtins utilities to separate Builtin module

mirryi committed:
Small code clarity fixes

mirryi committed:
Make ApBuiltin ident naming consistent in usage

cyrus- commented on src/hazelcore/Builtin.re:
```diff
@@ -0,0 +1,115 @@
+     Create a built-in function.
+   */
+  let mk = (ident: Var.t, ty: HTyp.t, fn: f): t => {
+    let eval = (args, evaluate) => fn(args, evaluate);
```
isn't this the same as saying let eval = f?

mirryi replied:
Oops, yeah. Left over from when I was errneously passing the unevaluated expression into fn.

cyrus- commented on src/hazelcore/Builtins.re:
```diff
@@ -0,0 +1,66 @@
+    | BoxedValue(FloatLit(f)) =>
+      let i = int_of_float(f);
+      BoxedValue(IntLit(i));
+    | BoxedValue(d1)
```
if its any boxed value other than a float lit it should raise an exception not return Indet (also in other cases in this file)

mirryi committed:
Clean up Builtin.mk

mirryi committed:
Raise exception on wrong BoxedValue in builtin evaluators

cyrus- commented on src/hazelcore/Builtin.re:
```diff
@@ -0,0 +1,113 @@
+module Impl = {
+  /* Evaluator alias. */
+  [@deriving sexp]
+  type evaluate = DHExp.t => EvaluatorResult.t;
```
this is used in the type `eval` above, maybe move this out to the top level / interface.

everything else in Impl you can also move out and just leave out of the rei interface to avoid it being exported

cyrus- reviewed (changes requested):
one last cleanup :)

mirryi committed:
Move items out of Builtin.Impl

mirryi committed:
Expose Builtin.mk in interface

mirryi commented:
Are these all the built-ins we want for now? Do we want to add/remove any?

cyrus- commented:
These should be fine for now -- I don't want to flood the context sidebar with builtins right now. I'm going to add "organize context hierarchically" as a starter project, but we don't need to do that to merge this mechanism in. It looks pretty straightforward to add new builtins now so adding more builtins as needed can be handled by other folks as needed.

cyrus- reviewed (approved):
Looks great now, thanks Eric!
